src/wp-includes/interactivity-api/class-wp-interactivity-api.php

```diff
@@ -167,10 +167,54 @@ public function print_client_interactivity_data() {
 		}
 
 		if ( ! empty( $interactivity_data ) ) {
+
+			/*
+			 * This data will be printed as JSON inside a script tag like this:
+			 *   <script type="application/json"></script>
+			 *
+			 * A script tag must be closed by a sequence beginning with `</`. It's impossible to
+			 * close a script tag without using `<`. We ensure that `<` is escaped and `/` can
+			 * remain unescaped, so `</script>` will be printed as `\u003C/script\u00E3`.
+			 *
+			 *   - JSON_HEX_TAG: All < and > are converted to \u003C and \u003E.
+			 *   - JSON_UNESCAPED_SLASHES: Don't escape /.
+			 *
+			 * @see https://www.php.net/manual/en/json.constants.php for details on these constants.
+			 * @see https://html.spec.whatwg.org/#script-data-state for details on script
+			 *      tag parsing.
+			 */
+			$json_encode_flags = JSON_HEX_TAG | JSON_UNESCAPED_SLASHES;
+			if ( 'UTF-8' === get_option( 'blog_charset' ) ) {
```

Comment:
https://github.com/WordPress/wordpress-develop/pull/6535

Reply:
Thanks! It looks like the `_canonical_charset()` function (intended to normalize to `"UTF-8"`) is applied as a filter by default, so maybe this is sufficient as is? Your proposed patch would make that filter more robust without any changes here:

https://github.com/WordPress/wordpress-develop/blob/12dc16563b2700b1ef60102a3bb4321298bb4559/src/wp-includes/default-filters.php#L292

If the encoding is UTF-8 and we fail to it, it will do little harm. The only issue is that valid unicode would be escaped to its `\u1234` form.